src/Catalyst.Core.Modules.Dfs/CoreApi/BlockApi.cs

```diff
@@ -0,0 +1,396 @@
+                Size = emptyDirectory.ToArray().Length
+            };
+        }
+
+        private static DataBlock GetEmptyNode()
+        {
+            var emptyNode = ObjectApi.GetEmptyNode();
+            return new DataBlock
+            {
+                DataBytes = emptyNode.ToArray(),
+                Id = emptyNode.Id,
+                Size = emptyNode.ToArray().Length
+            };
+        }
+
+        private FileStore<Cid, DataBlock> _store;
+
+        private readonly IDhtApi _dhtApi;
+        private readonly ISwarmApi _swarmApi;
+        private readonly IBitSwapApi _bitSwapApi;
+        private readonly DfsOptions _dfsOptions;
+        private readonly DfsState _dfsState;
+
+        public IPinApi PinApi { get; set; }
+
+        /// <summary>
+        /// </summary>
+        /// <param name="bitSwapApi"></param>
+        /// <param name="dhtApi"></param>
+        /// <param name="swarmApi"></param>
+        /// <param name="dfsOptions"></param>
+        public BlockApi(IBitSwapApi bitSwapApi,
+            IDhtApi dhtApi,
+            ISwarmApi swarmApi,
+            DfsState dfsState,
+            DfsOptions dfsOptions)
+        {
+            _bitSwapApi = bitSwapApi;
+            _dhtApi = dhtApi;
+            _swarmApi = swarmApi;
+            _dfsOptions = dfsOptions;
+            _dfsState = dfsState;
+        }
+
+        public FileStore<Cid, DataBlock> Store
+        {
+            get
+            {
+                if (_store != null)
+                {
+                    return _store;
+                }
+
+                var folder = Path.Combine(_dfsOptions.Repository.Folder, "blocks");
+
+                if (!Directory.Exists(folder))
+                {
+                    Directory.CreateDirectory(folder);
+                }
+
+                _store = new FileStore<Cid, DataBlock>
+                {
+                    Folder = folder,
+                    NameToKey = cid => cid.Hash.ToBase32(),
+                    KeyToName = key =>
+                        new MultiHash(key.FromBase32()), // @TODO assumes base32 should get from HashLibProvider prop
+
+                    Serialize = async (stream, cid, block, cancel) =>
+                    {
+                        await stream.WriteAsync(block.DataBytes, 0, block.DataBytes.Length, cancel)
+                           .ConfigureAwait(false);
+                    },
+
+                    Deserialize = async (stream, cid, cancel) =>
+                    {
+                        var block = new DataBlock
+                        {
+                            Id = cid,
+                            Size = stream.Length
+                        };
+
+                        block.DataBytes = new byte[block.Size];
+
+                        for (int i = 0, n; i < block.Size; i += n)
+                        {
+                            n = await stream.ReadAsync(block.DataBytes, i, (int) block.Size - i, cancel)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not update the loop counter 'n' within the loop body.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)